core/primitives/src/stateless_validation/contract_distribution.rs

```diff
@@ -0,0 +1,240 @@
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ContractCodeRequestV1 {
+    inner: ContractCodeRequestInner,
+    /// Signature of the inner.
+    signature: Signature,
+}
+
+impl ContractCodeRequestV1 {
+    fn new(
+        next_chunk: ChunkProductionKey,
+        contracts: Vec<CodeHash>,
+        signer: &ValidatorSigner,
+    ) -> Self {
+        let inner = ContractCodeRequestInner::new(next_chunk, contracts);
+        let signature = signer.sign_contract_code_request(&inner);
+        Self { inner, signature }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ContractCodeRequestInner {
+    /// Production metadata of the chunk created after the chunk the accesses belong to.
+    /// We associate this message with the next-chunk info because this message is generated
+    /// and distributed while generating the state-witness of the next chunk
+    /// (by the chunk producer of the next chunk).
+    // TODO(#11099): Consider simplifying this with the ChunkHash of the prev_chunk (the one the accesses belong to).
+    next_chunk: ChunkProductionKey,
+    /// List of code-hashes for the contracts accessed.
+    contracts: Vec<CodeHash>,
+    signature_differentiator: SignatureDifferentiator,
+}
+
+impl ContractCodeRequestInner {
+    fn new(next_chunk: ChunkProductionKey, contracts: Vec<CodeHash>) -> Self {
+        Self {
+            next_chunk,
+            contracts,
+            signature_differentiator: "ContractCodeRequestInner".to_owned(),
+        }
+    }
+}
+
+// Data structures for chunk producers to send contract code to chunk validators as response to ContractCodeRequest.
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ContractCodeResponse {
+    V1(ContractCodeResponseV1),
+}
+
+impl ContractCodeResponse {
+    pub fn new(contracts: &Vec<CodeBytes>, signer: &ValidatorSigner) -> Self {
+        Self::V1(ContractCodeResponseV1::new(contracts, signer))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ContractCodeResponseV1 {
+    inner: ContractCodeResponseInner,
+    /// Signature of the inner.
+    signature: Signature,
+}
+
+impl ContractCodeResponseV1 {
+    fn new(contracts: &Vec<CodeBytes>, signer: &ValidatorSigner) -> Self {
+        let inner = ContractCodeResponseInner::new(contracts);
+        let signature = signer.sign_contract_code_response(&inner);
+        Self { inner, signature }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ContractCodeResponseInner {
+    /// Code for the contracts.
+    compressed_contracts: CompressedContractCode,
+    /// Total size (in number of bytes) of the "uncompressed" form of contracts.
+    /// Used to limit the output while uncompressing it.
+    total_size: usize,
```

Comment:
`CompressedData` allows decoding (uncompressing) with a limit (other than the constant).

https://github.com/near/nearcore/blob/5c9b992323e69239df6e4621aa0045ed990dbb83/core/primitives/src/utils/compression.rs#L44

If we send the total size, we can use that as the limit while decoding instead of relying on a much higher upper-bound.

Reply:
but what exactly do we need that limit for? we cannot really rely on it as a security measure since this is external data and malicious node could set that to any value anyway

Reply:
yeah, does not make sense for security, the entire inner message is signed together anyways, it does not separate the limit and data. removing the size.